spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -286,22 +292,45 @@ protected void registerHandlerMethodOnListenedChannel(Method method, StreamListe
 					"An output channel must be specified for a method that can return a value");
 		}
 		StreamListenerMethodUtils.validateStreamListenerMessageHandler(method);
-		StreamListenerMessageHandler handler = new StreamListenerMessageHandler(invocableHandlerMethod);
-		handler.setApplicationContext(this.applicationContext);
-		handler.setChannelResolver(this.binderAwareChannelResolver);
-		if (!StringUtils.isEmpty(defaultOutputChannel)) {
-			handler.setOutputChannelName(defaultOutputChannel);
-		}
-		handler.afterPropertiesSet();
-		channel.subscribe(handler);
+		mappedListenerMethods.add(streamListener.value(),
+				new StreamListenerHandlerMethodMapping(bean, method, streamListener.condition(), defaultOutputChannel));
 	}
 
 	@Override
-	public void afterSingletonsInstantiated() {
-		// Dump the mappings after the context has been created, ensuring that beans can
-		// be processed correctly
-		// again.
-		this.mappedBindings.clear();
+	public final void afterSingletonsInstantiated() {
+		for (Map.Entry<String, List<StreamListenerHandlerMethodMapping>> mappedBindingEntry : mappedListenerMethods
+				.entrySet()) {
+			Collection<DispatchingStreamListenerMessageHandler.ConditionalStreamListenerHandler> handlerMethods = new ArrayList<>();
+			for (StreamListenerHandlerMethodMapping mapping : mappedBindingEntry.getValue()) {
+				final InvocableHandlerMethod invocableHandlerMethod = this.messageHandlerMethodFactory
+						.createInvocableHandlerMethod(mapping.getTargetBean(),
+								checkProxy(mapping.getMethod(), mapping.getTargetBean()));
+				StreamListenerMessageHandler streamListenerMessageHandler = new StreamListenerMessageHandler(
+						invocableHandlerMethod);
+				streamListenerMessageHandler.setApplicationContext(this.applicationContext);
+				streamListenerMessageHandler.setBeanFactory(this.applicationContext.getBeanFactory());
+				if (StringUtils.hasText(mapping.getDefaultOutputChannel())) {
+					streamListenerMessageHandler.setOutputChannelName(mapping.getDefaultOutputChannel());
+				}
+				streamListenerMessageHandler.afterPropertiesSet();
+				if (StringUtils.hasText(mapping.getCondition())) {
+					Expression condition = new SpelExpressionParser().parseExpression(mapping.getCondition());
```

Comment:
`SpelExpressionParser` is thread-safe and stateless. Can be just as `static final` property.